### PR TITLE
[1LP][RFR] Test vddk version(ssa)

### PR DESCRIPTION
**Pull request by rlbabyuk**
Purpose or Intent
===========
test different vddk version against different VMware provider versions

{{pytest: cfme/tests/infrastructure/test_ssa_vddk.py --long-running --use-provider complete}}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 5cad025fbfb41e7040bc7e7eb9d0854ccbde98e2:

`cfme/tests/infrastructure/test_ssa_vddk.py`:
- :bangbang: Line 99:13: **E116** *unexpected indentation (comment)*
- :heavy_exclamation_mark: Line 112:1: **W391** *blank line at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        provider.refresh_provider_relationships()
+    except Exception as e:
+        logger.exception(e)
+
+
+@pytest.mark.long_running
+def test_ssa_vddk(vm, configure_vddk):
+    """
+        Check if different version of vddk works with provider
+    """
```
Comment (mkoura) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: OPTIONAL: please use docstrings formatting consistently (preferably google-style dosctrings)

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+from cfme.utils.generators import random_vm_name
+from cfme.utils.log import logger
+from markers.env_markers.provider import ONE_PER_VERSION
+
+pytestmark = [
+    pytest.mark.tier(3),
+    test_requirements.smartstate,
+    pytest.mark.meta(server_roles="+smartproxy +smartstate"),
+    pytest.mark.provider([VMwareProvider],
+                         selector=ONE_PER_VERSION, scope='module'),
```
Comment (mkoura) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: REQUIRED: please follow our [style guide](http://cfme-tests.readthedocs.io/guides/dev_guide.html) and use one indent.

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    collected_files = []
+    for file in ["/etc/hosts", "/etc/passwd"]:
+        collected_files.append({"Name": file, "Collect Contents?": True})
+
+    analysis_profile_name = 'ssa_analysis_{}'.format(fauxfactory.gen_alphanumeric())
+    analysis_profile = AnalysisProfile(name=analysis_profile_name,
+                                       description=analysis_profile_name,
+                                       profile_type=AnalysisProfile.VM_TYPE,
+                                       categories=["System"],
+                                       files=collected_files)
```
Comment (mkoura) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: REQUIRED: please follow our [style guide](http://cfme-tests.readthedocs.io/guides/dev_guide.html) and use one indent.

**Comment by mkoura** [comment]
Please take a look at the PRT failures.

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    analysis_profile.create()
+    yield analysis_profile
+    analysis_profile.delete()
+
+
+@pytest.yield_fixture(params=vddk_versions, ids=([item for item in vddk_versions]), scope='module')
+def configure_vddk(request, appliance, provider, vm):
+    vddk_version = request.param
+    vddk_url = conf.cfme_data.get("basic_info").get("vddk_url").get(vddk_version)
+    host = vm.get_detail(properties=("Relationships", "Host"))
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: **REQUIRED**
```python
host = vm.entities.summary("Relationships").get_text_of("Host")
```

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+
+@pytest.mark.long_running
+def test_ssa_vddk(vm, configure_vddk):
+    """Check if different version of vddk works with provider
+
+    """
+    vm.smartstate_scan(wait_for_task_result=True)
+    c_users = vm.get_detail(properties=('Security', 'Users'))
+    c_groups = vm.get_detail(properties=('Security', 'Groups'))
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: **REQUIRED**
```python
c_users = vm.entities.summary('Security').get_text_of('Users')
c_groups = vm.entities.summary('Security').get_text_of('Groups')
```

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    assert any([c_users != 0, c_groups != 0])
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: **QUESTION**
Did you mean `all()`

Comment (rlbabyuk) [reply cfme/tests/infrastructure/test_ssa_vddk.py]: In this case, any() will do, as it is enough to check if ssa can grep some info, more detailed content testing is done in separate cases

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    analysis_profile.create()
+    yield analysis_profile
+    analysis_profile.delete()
+
+
+@pytest.yield_fixture(params=vddk_versions, ids=([item for item in vddk_versions]), scope='module')
+def configure_vddk(request, appliance, provider, vm):
+    vddk_version = request.param
+    vddk_url = conf.cfme_data.get("basic_info").get("vddk_url").get(vddk_version)
+    host = vm.entities.summary("Relationships").get_text_of("Host")
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: **REQUIRED**
```python
view = navigate_to(vm, "Details")
host = view.entities.summary("Relationships").get_text_of("Host")
```

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+
+@pytest.mark.long_running
+def test_ssa_vddk(vm, configure_vddk):
+    """Check if different version of vddk works with provider
+
+    """
+    vm.smartstate_scan(wait_for_task_result=True)
+    c_users = vm.entities.summary('Security').get_text_of('Users')
+    c_groups = vm.entities.summary('Security').get_text_of('Groups')
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: **REQUIRED**
```python
view = navigate_to(vm, "Details")
c_users = view.entities.summary('Security').get_text_of('Users')
c_groups = view.entities.summary('Security').get_text_of('Groups')
```

**Comment by dajoRH** [comment]
Lint report for commit 7069dd6c63b7463722b6e24809d6fb3e2eb5eb77:

`cfme/tests/infrastructure/test_ssa_vddk.py`:
- :bangbang: Line 6:1: **E303** *too many blank lines (3)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+def configure_vddk(request, appliance, provider, vm):
+    vddk_version = request.param
+    vddk_url = conf.cfme_data.get("basic_info").get("vddk_url").get(vddk_version)
+    view = navigate_to(vm, 'Details')
+    host = view.entities.summary("Relationships").get_text_of("Host")
+    setup_host_creds(provider, host)
+    appliance.install_vddk(vddk_url=vddk_url)
+    yield
+    appliance.uninstall_vddk()
+    setup_host_creds(provider, host, remove_creds=True)
```
Comment (jkrocil) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: If anything before `yield` fails, the finalizer part will not be executed.
If e.g. `appliance.install_vddk(vddk_url=vddk_url)` fails, `setup_host_creds(provider, host, remove_creds=True)` will never be executed so the host credentials (which have already been set) will never be unset.

We have to protect against anomalies like this one. Either the finalizer is important and we need to ensure that it is always executed or it is not important and we don't need it at all. Which one is it? :blush:

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+
+@pytest.yield_fixture(scope="module")
+def vm(provider, small_template_modscope, ssa_analysis_profile):
+    """ Fixture to provision instance on the provider """
+    vm_name = random_vm_name("ssa", max_length=16)
+    vm_obj = VM.factory(vm_name, provider, template_name=small_template_modscope.name)
+    vm_obj.create_on_provider(find_in_cfme=True, allow_skip="default")
+    provider.mgmt.start_vm(vm_obj.name)
+    provider.mgmt.wait_vm_running(vm_obj.name)
```
Comment (jkrocil) [inline cfme/tests/infrastructure/test_ssa_vddk.py]: Same here. If we are unable to wait for the vm to be running, the yield fixture fails. But the finalizer is never executed, so we never attempt to kill the vm...

File: cfme/tests/infrastructure/test_ssa_vddk.py (hunk 9 above)
Comment (rlbabyuk) [reply cfme/tests/infrastructure/test_ssa_vddk.py]: Added finalizer here as we need this clean up )

**Comment by dajoRH** [comment]
Lint report for commit 1595b5f20d1dc4e4f37398aa30d49a045eed3a68:

`cfme/tests/infrastructure/test_ssa_vddk.py`:
- :bangbang: Line 86:1: **E303** *too many blank lines (3)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/infrastructure/test_ssa_vddk.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                                       description=analysis_profile_name,
+                                       profile_type=AnalysisProfile.VM_TYPE,
+                                       categories=["System"],
+                                       files=collected_files)
+    analysis_profile.create()
+    yield analysis_profile
+    analysis_profile.delete()
+
+
+@pytest.fixture(params=vddk_versions, ids=([item for item in vddk_versions]), scope='module')
```
[the remark this answers is not in the document]
Comment (rlbabyuk) [reply cfme/tests/infrastructure/test_ssa_vddk.py]: Removed yield and add finalizer here instead, there is no need for yield_fixture here

File: cfme/tests/infrastructure/test_ssa_vddk.py (hunk 10 above)
Comment (rlbabyuk) [reply cfme/tests/infrastructure/test_ssa_vddk.py]: Added finalizer below

**Comment by rlbabyuk** [comment]
@psav, there were some changes, I've updated comments

**Pull request merged**